File: ddosshield.c
/*
 * ddosshield - OpenBSD DDoS detection and remote syslog reporter
 *
 * Periodically polls the pf(4) state table via pfctl(8), counts
 * connections per source IP, and forwards a RFC 3164 syslog message
 * over UDP to a remote syslog server whenever a source IP exceeds the
 * configured connection threshold within the sampling interval.
 *
 * Build:   make
 * Usage:   ddosshield [-v] [-c config] [-H host] [-p port]
 *                     [-t threshold] [-i interval]
 */

#include <sys/types.h>
#include <sys/socket.h>

#include <netinet/in.h>
#include <arpa/inet.h>

#include <err.h>
#include <errno.h>
#include <netdb.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define PROGNAME		"ddosshield"
#define DEFAULT_CONF		"/etc/ddosshield.conf"
#define PFCTL_CMD		"pfctl -s state 2>/dev/null"

/* Remote syslog defaults */
#define DEFAULT_PORT		514
#define DEFAULT_THRESHOLD	100	/* connections per interval */
#define DEFAULT_INTERVAL	10	/* seconds */

/* RFC 3164: facility LOCAL0 (16), severity WARNING (4) */
#define SYSLOG_PRI		((16 * 8) + 4)

/* Hash table sizing */
#define HASH_SIZE		4096

#define MAX_ADDR_LEN		46	/* max IPv6 string length + NUL */
#define MAX_LINE_LEN		512
#define MAX_HOST_LEN		256

/* Per-source-IP tracking entry */
struct ip_entry {
	char		 addr[MAX_ADDR_LEN];
	int		 count;
	int		 alerted;
	struct ip_entry	*next;
};

/* Runtime configuration */
static struct {
	char	remote_host[MAX_HOST_LEN];
	int	remote_port;
	int	threshold;
	int	interval;
	int	verbose;
} cfg = {
	.remote_host	= "",
	.remote_port	= DEFAULT_PORT,
	.threshold	= DEFAULT_THRESHOLD,
	.interval	= DEFAULT_INTERVAL,
	.verbose	= 0,
};

static int			 sockfd = -1;
static struct sockaddr_in	 remote_sa;
static volatile sig_atomic_t	 running = 1;

/* Hash table of source IP entries, reset each interval */
static struct ip_entry		*iptable[HASH_SIZE];

static void
handle_signal(int sig)
{
	(void)sig;
	running = 0;
}

/*
 * Send an RFC 3164 syslog message to the remote server over UDP.
 * Format: <PRI>TIMESTAMP HOSTNAME TAG: MESSAGE
 */
static int
remote_syslog(const char *fmt, ...)
{
	char		 msg[512];
	char		 buf[1024];
	char		 ts[32];
	char		 hostname[256];
	va_list		 ap;
	time_t		 now;
	struct tm	*tm;
	int		 n;

	va_start(ap, fmt);
	vsnprintf(msg, sizeof(msg), fmt, ap);
	va_end(ap);

	now = time(NULL);
	tm = localtime(&now);
	strftime(ts, sizeof(ts), "%b %e %H:%M:%S", tm);

	if (gethostname(hostname, sizeof(hostname)) == -1)
		strlcpy(hostname, "localhost", sizeof(hostname));
	hostname[sizeof(hostname) - 1] = '\0';

	n = snprintf(buf, sizeof(buf), "<%d>%s %s %s: %s",
	    SYSLOG_PRI, ts, hostname, PROGNAME, msg);
	if (n < 0)
		return -1;
	if (n >= (int)sizeof(buf))
		n = sizeof(buf) - 1;

	if (sendto(sockfd, buf, (size_t)n, 0,
	    (struct sockaddr *)&remote_sa, sizeof(remote_sa)) == -1) {
		if (cfg.verbose)
			warn("sendto remote syslog");
		return -1;
	}
	return 0;
}

/* djb2-based hash for an IP address string */
static unsigned int
hash_addr(const char *addr)
{
	unsigned int	 h = 5381;
	int		 c;

	while ((c = (unsigned char)*addr++) != '\0')
		h = ((h << 5) + h) ^ (unsigned int)c;
	return h % HASH_SIZE;
}

/* Look up or allocate an ip_entry for addr */
static struct ip_entry *
ip_entry_get(const char *addr)
{
	unsigned int	 h = hash_addr(addr);
	struct ip_entry	*e;

	for (e = iptable[h]; e != NULL; e = e->next) {
		if (strcmp(e->addr, addr) == 0)
			return e;
	}

	e = calloc(1, sizeof(*e));
	if (e == NULL)
		return NULL;
	strlcpy(e->addr, addr, sizeof(e->addr));
	e->next = iptable[h];
	iptable[h] = e;
	return e;
}

/* Free all entries in the hash table */
static void
iptable_reset(void)
{
	struct ip_entry	*e, *next;
	int		 i;

	for (i = 0; i < HASH_SIZE; i++) {
		for (e = iptable[i]; e != NULL; e = next) {
			next = e->next;
			free(e);
		}
		iptable[i] = NULL;
	}
}

/*
 * Extract the source IP address from a pfctl -s state output line.
 *
 * Line format (columns separated by whitespace):
 *   <iface> <proto> <src_addr:port> -> <dst_addr:port> -> <state>
 *
 * IPv4 src: "192.0.2.1:12345"   → extract everything before last ':'
 * IPv6 src: "[2001:db8::1]:12345" → strip brackets and port
 *
 * Returns 0 on success with addr filled in, -1 if line cannot be parsed.
 */
static int
parse_src_addr(const char *line, char *addr, size_t addrlen)
{
	const char	*p = line;
	const char	*field, *end, *colon;
	size_t		 flen;
	int		 i;

	/* Skip the first two whitespace-separated fields (iface, proto) */
	for (i = 0; i < 2; i++) {
		while (*p == ' ' || *p == '\t')
			p++;
		if (*p == '\0')
			return -1;
		while (*p != ' ' && *p != '\t' && *p != '\0')
			p++;
	}

	/* Skip leading whitespace before the source field */
	while (*p == ' ' || *p == '\t')
		p++;
	if (*p == '\0')
		return -1;

	field = p;
	/* The source field ends at the next whitespace */
	end = field;
	while (*end != ' ' && *end != '\t' && *end != '\0')
		end++;
	flen = (size_t)(end - field);

	if (flen == 0 || flen >= MAX_LINE_LEN)
		return -1;

	if (field[0] == '[') {
		/* IPv6: [addr]:port – strip brackets and port */
		const char *bracket_close = memchr(field, ']', flen);
		if (bracket_close == NULL)
			return -1;
		size_t ilen = (size_t)(bracket_close - field) - 1;
		if (ilen == 0 || ilen >= addrlen)
			return -1;
		memcpy(addr, field + 1, ilen);
		addr[ilen] = '\0';
	} else {
		/* IPv4: addr:port – everything before the last colon */
		char tmp[MAX_LINE_LEN];
		if (flen >= sizeof(tmp))
			return -1;
		memcpy(tmp, field, flen);
		tmp[flen] = '\0';
		colon = strrchr(tmp, ':');
		if (colon == NULL)
			return -1;
		size_t ilen = (size_t)(colon - tmp);
		if (ilen == 0 || ilen >= addrlen)
			return -1;
		memcpy(addr, tmp, ilen);
		addr[ilen] = '\0';
	}

	return 0;
}

/* Poll pfctl state table and check for threshold violations */
static void
check_states(void)
{
	FILE		*fp;
	char		 line[MAX_LINE_LEN];
	char		 src[MAX_ADDR_LEN];
	struct ip_entry	*e;

	fp = popen(PFCTL_CMD, "r");
	if (fp == NULL) {
		warn("popen: " PFCTL_CMD);
		return;
	}

	while (fgets(line, sizeof(line), fp) != NULL) {
		if (parse_src_addr(line, src, sizeof(src)) == -1)
			continue;

		/* Skip loopback addresses */
		if (strncmp(src, "127.", 4) == 0 || strcmp(src, "::1") == 0)
			continue;

		e = ip_entry_get(src);
		if (e == NULL)
			continue;
		e->count++;

		if (!e->alerted && e->count >= cfg.threshold) {
			const char *msg_fmt =
			    "Suspected DDoS attack from %s: "
			    "%d connections within %d seconds";

			if (cfg.verbose)
				fprintf(stderr, PROGNAME ": " msg_fmt "\n",
				    src, e->count, cfg.interval);

			remote_syslog(msg_fmt, src, e->count, cfg.interval);
			e->alerted = 1;
		}
	}
	pclose(fp);
}

/* Parse a simple "key value" configuration file */
static void
parse_config(const char *path)
{
	FILE	*fp;
	char	 line[256];
	char	 key[64], val[192];

	fp = fopen(path, "r");
	if (fp == NULL) {
		if (errno != ENOENT)
			warn("fopen: %s", path);
		return;
	}

	while (fgets(line, sizeof(line), fp) != NULL) {
		/* Skip blank lines and comments */
		if (line[0] == '#' || line[0] == '\n' || line[0] == '\r')
			continue;
		if (sscanf(line, "%63s %191s", key, val) != 2)
			continue;
		if (strcmp(key, "remote_host") == 0)
			strlcpy(cfg.remote_host, val, sizeof(cfg.remote_host));
		else if (strcmp(key, "remote_port") == 0)
			cfg.remote_port = atoi(val);
		else if (strcmp(key, "threshold") == 0)
			cfg.threshold = atoi(val);
		else if (strcmp(key, "interval") == 0)
			cfg.interval = atoi(val);
	}
	fclose(fp);
}

static void __dead
usage(void)
{
	fprintf(stderr,
	    "usage: " PROGNAME " [-v] [-c config] [-H host] [-p port]"
	    " [-t threshold] [-i interval]\n");
	exit(1);
}

int
main(int argc, char *argv[])
{
	const char	*conffile = DEFAULT_CONF;
	struct hostent	*he;
	int		 ch;

	/* Load config file first; CLI flags override below */
	parse_config(conffile);

	while ((ch = getopt(argc, argv, "vc:H:p:t:i:")) != -1) {
		switch (ch) {
		case 'v':
			cfg.verbose = 1;
			break;
		case 'c':
			conffile = optarg;
			parse_config(conffile);
			break;
		case 'H':
			strlcpy(cfg.remote_host, optarg,
			    sizeof(cfg.remote_host));
			break;
		case 'p':
			cfg.remote_port = atoi(optarg);
			break;
		case 't':
			cfg.threshold = atoi(optarg);
			break;
		case 'i':
			cfg.interval = atoi(optarg);
			break;
		default:
			usage();
		}
	}

	if (cfg.remote_host[0] == '\0')
		errx(1, "remote_host not configured (use -H or set in %s)",
		    conffile);
	if (cfg.threshold <= 0)
		errx(1, "threshold must be a positive integer");
	if (cfg.interval <= 0)
		errx(1, "interval must be a positive integer");

	/* Resolve remote syslog host before pledging */
	he = gethostbyname(cfg.remote_host);
	if (he == NULL)
		errx(1, "cannot resolve host: %s", cfg.remote_host);

	memset(&remote_sa, 0, sizeof(remote_sa));
	remote_sa.sin_family = AF_INET;
	remote_sa.sin_port = htons((uint16_t)cfg.remote_port);
	memcpy(&remote_sa.sin_addr, he->h_addr_list[0],
	    (size_t)he->h_length);

	/* Open UDP socket for remote syslog */
	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
	if (sockfd == -1)
		err(1, "socket");

	signal(SIGTERM, handle_signal);
	signal(SIGINT,  handle_signal);

	/*
	 * Restrict privileges with pledge(2).
	 * "proc exec" covers popen/pclose; "inet" covers UDP sendto;
	 * "stdio" covers file I/O and time.
	 */
#ifdef __OpenBSD__
	if (pledge("stdio proc exec inet", NULL) == -1)
		err(1, "pledge");
#endif

	if (cfg.verbose)
		fprintf(stderr,
		    PROGNAME ": started — remote=%s:%d threshold=%d "
		    "interval=%ds\n",
		    cfg.remote_host, cfg.remote_port,
		    cfg.threshold, cfg.interval);

	while (running) {
		check_states();
		sleep((unsigned int)cfg.interval);
		iptable_reset();
	}

	close(sockfd);
	iptable_reset();
	return 0;
}

File: ddosshield.m
#import <Foundation/Foundation.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <signal.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>

/* RFC 3164 syslog: LOCAL0 facility (16), WARNING severity (4) */
#define SYSLOG_PRI ((16 * 8) + 4)

static volatile int running = 1;

static void
handle_signal(int sig)
{
    (void)sig;
    running = 0;
}

/* ------------------------------------------------------------------ */
/* DDOSShield                                                           */
/* ------------------------------------------------------------------ */

@interface DDOSShield : NSObject {
    NSString            *_remoteHost;
    int                  _remotePort;
    int                  _threshold;
    int                  _interval;
    BOOL                 _verbose;
    int                  _sockfd;
    struct sockaddr_in   _remoteSA;
    NSMutableDictionary *_ipCounts;   /* NSString -> NSNumber (count)  */
    NSMutableDictionary *_ipAlerted;  /* NSString -> NSNumber (BOOL)   */
}

- (id)init;
- (void)dealloc;

/* Accessors */
- (NSString *)remoteHost;
- (void)setRemoteHost:(NSString *)host;
- (int)remotePort;
- (void)setRemotePort:(int)port;
- (int)threshold;
- (void)setThreshold:(int)t;
- (int)interval;
- (void)setInterval:(int)i;
- (BOOL)verbose;
- (void)setVerbose:(BOOL)v;

/* Operations */
- (void)loadConfig:(NSString *)path;
- (BOOL)setupSocket;
- (void)run;

@end

@implementation DDOSShield

- (id)init
{
    self = [super init];
    if (self) {
        _remoteHost = nil;
        _remotePort = 514;
        _threshold  = 100;
        _interval   = 10;
        _verbose    = NO;
        _sockfd     = -1;
        _ipCounts   = [[NSMutableDictionary alloc] init];
        _ipAlerted  = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)dealloc
{
    [_remoteHost release];
    [_ipCounts   release];
    [_ipAlerted  release];
    if (_sockfd != -1)
        close(_sockfd);
    [super dealloc];
}

/* --- Accessors ---------------------------------------------------- */

- (NSString *)remoteHost              { return _remoteHost; }
- (void)setRemoteHost:(NSString *)host
{
    [host retain];
    [_remoteHost release];
    _remoteHost = host;
}
- (int)remotePort                     { return _remotePort; }
- (void)setRemotePort:(int)port       { _remotePort = port; }
- (int)threshold                      { return _threshold; }
- (void)setThreshold:(int)t           { _threshold = t; }
- (int)interval                       { return _interval; }
- (void)setInterval:(int)i            { _interval = i; }
- (BOOL)verbose                       { return _verbose; }
- (void)setVerbose:(BOOL)v            { _verbose = v; }

/* --- Configuration ------------------------------------------------ */

/*
 * Parse a simple "key value" configuration file.
 * Lines beginning with '#' and blank lines are ignored.
 */
- (void)loadConfig:(NSString *)path
{
    NSString *content = [NSString stringWithContentsOfFile:path
                                                  encoding:NSUTF8StringEncoding
                                                     error:nil];
    if (!content)
        return;

    NSArray      *lines = [content componentsSeparatedByString:@"\n"];
    NSEnumerator *en    = [lines objectEnumerator];
    NSString     *line;

    while ((line = [en nextObject]) != nil) {
        NSString *trimmed = [line stringByTrimmingCharactersInSet:
            [NSCharacterSet whitespaceAndNewlineCharacterSet]];
        if ([trimmed length] == 0 || [trimmed hasPrefix:@"#"])
            continue;

        /* Split on any whitespace, skip empty tokens */
        NSArray        *parts  = [trimmed componentsSeparatedByCharactersInSet:
            [NSCharacterSet whitespaceAndNewlineCharacterSet]];
        NSMutableArray *tokens = [NSMutableArray array];
        NSEnumerator   *pen    = [parts objectEnumerator];
        NSString       *p;
        while ((p = [pen nextObject]) != nil) {
            if ([p length] > 0)
                [tokens addObject:p];
        }
        if ([tokens count] < 2)
            continue;

        NSString *key = [tokens objectAtIndex:0];
        NSString *val = [tokens objectAtIndex:1];

        if ([key isEqualToString:@"remote_host"])
            [self setRemoteHost:val];
        else if ([key isEqualToString:@"remote_port"])
            _remotePort = [val intValue];
        else if ([key isEqualToString:@"threshold"])
            _threshold = [val intValue];
        else if ([key isEqualToString:@"interval"])
            _interval = [val intValue];
    }
}

/* --- Networking --------------------------------------------------- */

/* Resolve the remote host and open the UDP socket. */
- (BOOL)setupSocket
{
    struct addrinfo  hints, *res;
    int              err;

    memset(&hints, 0, sizeof(hints));
    hints.ai_family   = AF_INET;
    hints.ai_socktype = SOCK_DGRAM;

    err = getaddrinfo([_remoteHost UTF8String], NULL, &hints, &res);
    if (err != 0) {
        NSLog(@"ddosshield: cannot resolve host %@: %s",
            _remoteHost, gai_strerror(err));
        return NO;
    }

    memset(&_remoteSA, 0, sizeof(_remoteSA));
    _remoteSA.sin_family = AF_INET;
    _remoteSA.sin_port   = htons((uint16_t)_remotePort);
    _remoteSA.sin_addr   = ((struct sockaddr_in *)res->ai_addr)->sin_addr;
    freeaddrinfo(res);

    _sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (_sockfd == -1) {
        NSLog(@"ddosshield: socket: %s", strerror(errno));
        return NO;
    }
    return YES;
}

/*
 * Send an RFC 3164 syslog message to the remote server over UDP.
 * Format: <PRI>TIMESTAMP HOSTNAME TAG: MESSAGE
 */
- (void)sendSyslog:(NSString *)message
{
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

    time_t    now = time(NULL);
    struct tm *tm = localtime(&now);
    char       tsbuf[32];
    strftime(tsbuf, sizeof(tsbuf), "%b %e %H:%M:%S", tm);

    NSString *hostname = [[NSProcessInfo processInfo] hostName];
    NSString *packet   = [NSString stringWithFormat:@"<%d>%s %@ ddosshield: %@",
        SYSLOG_PRI, tsbuf, hostname, message];

    const char *buf = [packet UTF8String];
    size_t      len = strlen(buf);

    if (sendto(_sockfd, buf, len, 0,
               (struct sockaddr *)&_remoteSA, sizeof(_remoteSA)) == -1) {
        if (_verbose)
            NSLog(@"ddosshield: sendto: %s", strerror(errno));
    }

    [pool drain];
}

/* --- pf state parsing --------------------------------------------- */

/*
 * Extract the source IP address from a pfctl(8) -s state line.
 *
 * Line format (whitespace-separated):
 *   <iface> <proto> <src_addr:port> -> <dst_addr:port> -> <state>
 *
 * IPv4 src:  "192.0.2.1:12345"     → strip last ':' and port
 * IPv6 src:  "[2001:db8::1]:12345" → strip brackets and port
 *
 * Returns nil when the line cannot be parsed.
 */
- (NSString *)parseSrcFromLine:(NSString *)line
{
    NSArray        *fields = [line componentsSeparatedByCharactersInSet:
        [NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSMutableArray *tokens = [NSMutableArray array];
    NSEnumerator   *en     = [fields objectEnumerator];
    NSString       *f;

    while ((f = [en nextObject]) != nil) {
        if ([f length] > 0)
            [tokens addObject:f];
    }

    /* Minimum viable line: iface proto src */
    if ([tokens count] < 3)
        return nil;

    NSString *src = [tokens objectAtIndex:2];

    /* IPv6: [addr]:port */
    if ([src hasPrefix:@"["]) {
        NSRange close = [src rangeOfString:@"]"];
        if (close.location == NSNotFound)
            return nil;
        return [src substringWithRange:NSMakeRange(1, close.location - 1)];
    }

    /* IPv4: addr:port — everything before the last colon */
    NSRange lastColon = [src rangeOfString:@":" options:NSBackwardsSearch];
    if (lastColon.location == NSNotFound)
        return nil;
    return [src substringToIndex:lastColon.location];
}

/* --- Monitoring --------------------------------------------------- */

/*
 * Poll the pf(4) state table via pfctl(8) and check each source IP
 * against the configured threshold, sending a remote syslog alert for
 * any IP that has exceeded it during this interval.
 */
- (void)checkStates
{
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

    NSTask *task = [[NSTask alloc] init];
    [task setLaunchPath:@"/sbin/pfctl"];
    [task setArguments:[NSArray arrayWithObjects:@"-s", @"state", nil]];

    NSPipe *outPipe = [NSPipe pipe];
    [task setStandardOutput:outPipe];
    [task setStandardError:[NSFileHandle fileHandleWithNullDevice]];
    [task launch];

    NSData *data = [[outPipe fileHandleForReading] readDataToEndOfFile];
    [task waitUntilExit];
    [task release];

    NSString *output = [[[NSString alloc] initWithData:data
                                              encoding:NSUTF8StringEncoding]
                        autorelease];
    if (!output) {
        [pool drain];
        return;
    }

    NSArray      *lines = [output componentsSeparatedByString:@"\n"];
    NSEnumerator *en    = [lines objectEnumerator];
    NSString     *line;

    while ((line = [en nextObject]) != nil) {
        NSString *src = [self parseSrcFromLine:line];
        if (!src)
            continue;

        /* Skip loopback addresses */
        if ([src hasPrefix:@"127."] || [src isEqualToString:@"::1"])
            continue;

        NSNumber *count    = [_ipCounts objectForKey:src];
        int       newCount = count ? [count intValue] + 1 : 1;
        [_ipCounts setObject:[NSNumber numberWithInt:newCount] forKey:src];

        NSNumber *alerted = [_ipAlerted objectForKey:src];
        if (!alerted && newCount >= _threshold) {
            NSString *msg = [NSString stringWithFormat:
                @"Suspected DDoS attack from %@: "
                "%d connections within %d seconds",
                src, newCount, _interval];

            if (_verbose)
                NSLog(@"ddosshield: %@", msg);

            [self sendSyslog:msg];
            [_ipAlerted setObject:[NSNumber numberWithBool:YES] forKey:src];
        }
    }

    [pool drain];
}

/* Reset per-IP counters and alert flags at the end of each interval. */
- (void)resetTable
{
    [_ipCounts  removeAllObjects];
    [_ipAlerted removeAllObjects];
}

/* Main monitoring loop. */
- (void)run
{
    while (running) {
        [self checkStates];
        sleep((unsigned int)_interval);
        [self resetTable];
    }
}

@end

/* ------------------------------------------------------------------ */
/* Entry point                                                          */
/* ------------------------------------------------------------------ */

static void
usage(const char *prog)
{
    fprintf(stderr,
        "usage: %s [-v] [-c config] [-H host] [-p port]"
        " [-t threshold] [-i interval]\n",
        prog);
    exit(1);
}

int
main(int argc, const char *argv[])
{
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];

    DDOSShield *shield   = [[[DDOSShield alloc] init] autorelease];
    const char *conffile = "/etc/ddosshield.conf";

    /*
     * Two-pass argument handling:
     *   Pass 1 – scan for -c to find the config file to load.
     *   Pass 2 – getopt processes all flags; CLI values override config.
     * This ensures that -H/-p/-t/-i always win over config values
     * regardless of where -c appears on the command line.
     */

    /* Pass 1: locate -c */
    for (int i = 1; i < argc - 1; i++) {
        if (strcmp(argv[i], "-c") == 0) {
            conffile = argv[i + 1];
            break;
        }
    }
    [shield loadConfig:[NSString stringWithUTF8String:conffile]];

    /* Pass 2: apply all CLI flags (override any config-file values) */
    int ch;
    while ((ch = getopt(argc, (char *const *)argv, "vc:H:p:t:i:")) != -1) {
        switch (ch) {
        case 'v':
            [shield setVerbose:YES];
            break;
        case 'c':
            /* already handled in pass 1 */
            break;
        case 'H':
            [shield setRemoteHost:[NSString stringWithUTF8String:optarg]];
            break;
        case 'p':
            [shield setRemotePort:atoi(optarg)];
            break;
        case 't':
            [shield setThreshold:atoi(optarg)];
            break;
        case 'i':
            [shield setInterval:atoi(optarg)];
            break;
        default:
            usage(argv[0]);
        }
    }

    if (![shield remoteHost] || [[shield remoteHost] length] == 0) {
        fprintf(stderr,
            "ddosshield: remote_host not configured "
            "(use -H or set in %s)\n", conffile);
        exit(1);
    }
    if ([shield threshold] <= 0) {
        fprintf(stderr, "ddosshield: threshold must be a positive integer\n");
        exit(1);
    }
    if ([shield interval] <= 0) {
        fprintf(stderr, "ddosshield: interval must be a positive integer\n");
        exit(1);
    }

    if (![shield setupSocket])
        exit(1);

    signal(SIGTERM, handle_signal);
    signal(SIGINT,  handle_signal);

    /*
     * Restrict privileges with pledge(2) on OpenBSD.
     * "proc exec" is needed for NSTask; "inet" for UDP sendto.
     */
#ifdef __OpenBSD__
    if (pledge("stdio proc exec inet", NULL) == -1) {
        NSLog(@"ddosshield: pledge: %s", strerror(errno));
        exit(1);
    }
#endif

    if ([shield verbose])
        NSLog(@"ddosshield: started — remote=%@:%d threshold=%d interval=%ds",
            [shield remoteHost], [shield remotePort],
            [shield threshold], [shield interval]);

    [shield run];

    [pool drain];
    return 0;
}
